### PR TITLE
Subgraph support for quantization tools

### PR DESCRIPTION
**Pull request by zhanghuanrong**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/python/tools/quantization/onnx_quantizer.py [hunk 1]
```diff
@@ -28,8 +28,10 @@ class ONNXQuantizer:
     def __init__(self, model, per_channel, reduce_range, mode, static, weight_qType, input_qType, tensors_range,
                  nodes_to_quantize, nodes_to_exclude, op_types_to_quantize, extra_options={}):
 
-        # run shape inference on the model
-        model = onnx.shape_inference.infer_shapes(model)
+        # run shape inference on the model (enabled by default)
+        self.extra_options = extra_options if extra_options is not None else {}
+        if not ('disableShapeInference' in self.extra_options and self.extra_options['disableShapeInference']):
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/onnx_quantizer.py:33]: could you please this option into the doc and comment in quantize.py

**Comment by david-macleod** [comment]
This looks great, I was looking to fix the same issue. Are you targeting the `1.9` release?

File: onnxruntime/python/tools/quantization/onnx_quantizer.py (hunk 1 above)
Comment (zhanghuanrong) [reply onnxruntime/python/tools/quantization/onnx_quantizer.py:32]: Added comment in the quantize.py

**Comment by zhanghuanrong** [comment]
> This looks great, I was looking to fix the same issue. Are you targeting the `1.9` release?

Should be in master branch soon.

File: onnxruntime/python/tools/quantization/onnx_quantizer.py (hunk 1 above)
Comment (yufenglee) [reply onnxruntime/python/tools/quantization/onnx_quantizer.py:32]: nit: use capital to keep consistent with other options

File: onnxruntime/python/tools/quantization/onnx_quantizer.py [hunk 2]
```diff
@@ -85,6 +94,55 @@ def __init__(self, model, per_channel, reduce_range, mode, static, weight_qType,
...
+                                      self.op_types_to_quantize,
+                                      self.extra_options)
+        sub_quanitzer.parent = self
+        sub_quanitzer.graph_scope = "{}{}/".format(self.graph_scope, graph_key)
+        sub_quanitzer.quantize_model()
+        return sub_quanitzer.model.model.graph
+
+    def quantize_node_with_sub_graph(self, node):
+        '''
+        Check subgraph, if any, quantize it and repleace it.
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/onnx_quantizer.py:125]: >repleace [](http://example.com/codeflow?start=48&length=8)

nit: replace

File: onnxruntime/python/tools/quantization/onnx_quantizer.py [hunk 3]
```diff
@@ -85,6 +94,55 @@ def __init__(self, model, per_channel, reduce_range, mode, static, weight_qType,
...
+        return new_nodes added for quantizing subgraph
+        '''
+        graph_attrs = [attr for attr in node.attribute if attr.type == 5 or attr.type == 10]
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/onnx_quantizer.py:128]: >5 [](http://example.com/codeflow?start=71&length=1)

nit: AttributeProto.GRAPH and AttributeProto.GRAPHS

File: onnxruntime/python/tools/quantization/quantize.py [hunk 4]
```diff
@@ -184,6 +184,11 @@ def quantize_static(model_input,
             extra.Sigmoid.nnapi = True/False  (Default is False)
             ActivationSymmetric = True/False: symmetrize calibration data for activations (default is False).
             WeightSymmetric = True/False: symmetrize calibration data for weights (default is True).
+            EnableSubgraph = True/False : Default is False. If enabled, subgraph will be quantized.
+                                          Dyanmic mode currently is supported. Will support more in future.
+            disableShapeInference = True/False : in dynamic quantize mode, shape inference is not must have
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/quantize.py:189]: >disableShapeInference [](http://example.com/codeflow?start=12&length=21)

nit: start with captial char

Comment (zhanghuanrong) [reply onnxruntime/python/tools/quantization/quantize.py]: fixed

File: onnxruntime/python/tools/quantization/onnx_quantizer.py [hunk 5]
```diff
@@ -85,6 +94,55 @@ def __init__(self, model, per_channel, reduce_range, mode, static, weight_qType,
...
+        if len(graph_attrs) == 0:
+            return node
+        node_name = node.name if node.name != "" else "{}_node_count_{}".format(node.op_type, len(self.new_nodes))
+        kwargs = {}
+        for attr in node.attribute:
+            if attr.type == 5:
```
[the remark this answers is not in the document]
Comment (zhanghuanrong) [reply onnxruntime/python/tools/quantization/onnx_quantizer.py]: changed

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**